### PR TITLE
Script to sync event topic

### PR DESCRIPTION
**Pull request by kruai**
# Overview

This PR is being created to address [ESSNTL-1281](https://issues.redhat.com/browse/ESSNTL-1281).
Context in this thread: https://ansible.slack.com/archives/CADT032B0/p1632242392478200
This goes through the messages in the `platform.inventory.events topic` and syncs it with the HBI DB. This is to help get Cyndi sync'd again.

## PR Checklist

- [x] Keep PR title short, ideally under 72 characters
- [x] Descriptive comments provided in complex code blocks
- [ ] Tests: validate optimal/expected output
- [ ] Tests: validate exceptions and failure scenarios
- [ ] Tests: edge cases
- [ ] Documentation, if this PR changes the way other services interact with host inventory
- [ ] Links to related PRs

## Secure Coding Practices Documentation Reference

You can find documentation on this checklist [here](https://github.com/RedHatInsights/secure-coding-checklist).

## Secure Coding Checklist

- [ ] Input Validation
- [ ] Output Encoding
- [ ] Authentication and Password Management
- [ ] Session Management
- [ ] Access Control
- [ ] Cryptographic Practices
- [ ] Error Handling and Logging
- [ ] Data Protection
- [ ] Communication Security
- [ ] System Configuration
- [ ] Database Security
- [ ] File Management
- [ ] Memory Management
- [ ] General Coding Practices

### CODE REVIEW

File: app/queue/queue.py [hunk 1]
```diff
@@ -161,6 +166,32 @@ def parse_operation_message(message):
...
+        # If the host exists in the DB, delete it.
+        if query.count():
+            for host_id, deleted in delete_hosts(query, event_producer, 1, shutdown_handler.shut_down):
+                if deleted:
+                    log_host_delete_succeeded(logger, host_id, "TOPIC-REBUILD")
+                else:
+                    log_host_delete_failed(logger, host_id, "TOPIC-REBUILD")
+    else:
+        query = session.query(Host).filter(
+            (Host.account == message["host"]["account"]) & (Host.id == UUID(message["host"]["id"]))
```
Comment (ryandillinfelton) [inline app/queue/queue.py:181]: What's up with the different indices for accessing the account number here and on line 171
```
message["account"]
```
vs
```
message["host"]["account"]
```
Is there a difference between `delete` messages and `update/create` message?

Comment (kruai) [reply app/queue/queue.py:181]: Yes, it's in our docs. `updated` event: https://consoledot.pages.redhat.com/docs/dev/services/inventory.html#_updated_event
`delete` event: https://consoledot.pages.redhat.com/docs/dev/services/inventory.html#_delete_event

**Comment by ryandillinfelton** [comment]
Okay, very rambly but I'm leaving this as a note to investigate tomorrow at least.

Could the discrepancy have been caused by create/update events that were put on the topic during normal inventory operation that didn't get read by the connector. If a duplicate host was updated right after cyndi shut down, and the update was added to the event topic, and then we ran the dupe deleting script, producing a delete event for the same host and deleting it from the inventory DB then:
- Would cyndi consider there to be a discrepancy even though, if processed chronologically it should see the current state of the host as deleted?
- Will this script correctly handle that case? 

This script looks to run chronologically, so it may do extra work recreating the deleted host and then deleting it again, but I think it would get it. I don't know how cyndi decides if the topic is in sync with the DB though so I can't speak to that.

Beyond that, if we know they are out of sync, but don't know why, how can we know that syncing them like this won't cause other problems with the data. We're just assuming that every delete on the topic is correct, and that every host missing in the database is correct. I'd like to get to the bottom of the cause, or at least better understand the possibilities before we run this.

**Comment by kruai** [comment]
> Okay, very rambly but I'm leaving this as a note to investigate tomorrow at least.
> 
I'll respond inline!
> Could the discrepancy have been caused by create/update events that were put on the topic during normal inventory operation that didn't get read by the connector. If a duplicate host was updated right after cyndi shut down, and the update was added to the event topic, and then we ran the dupe deleting script, producing a delete event for the same host and deleting it from the inventory DB then:
> 
I'm trying to follow, but I'm not fully sure I understand the scenario you're describing. Let me know if this is in line with what you mean:

1. Cyndi shuts down.
2. HBI consumes an `add_host` message from MQ, updates the host in the DB, and emits the "updated" event.
3. We run `host_delete_duplicates.py`, which removes the host from the DB, and emits a "delete" event.
4. We run `rebuild_events_topic.py`.

> * Would cyndi consider there to be a discrepancy even though, if processed chronologically it should see the current state of the host as deleted?
AFAIK, if all of those events were correctly added to `platform.inventory.events` and none were lost, there shouldn't be a discrepancy.
> * Will this script correctly handle that case?
If the host was removed from the DB and the "delete" event was produced, the record is considered to be in-sync. So, the script wouldn't do anything for that host.
> 
> This script looks to run chronologically, so it may do extra work recreating the deleted host and then deleting it again, but I think it would get it. I don't know how cyndi decides if the topic is in sync with the DB though so I can't speak to that.
I don't know Cyndi's logic either, beyond the basics. This script won't create new hosts or produce `created`/`updated` events, though; it just deletes hosts from the DB or adds a delete event to the MQ if a record is out of sync.
> 
> Beyond that, if we know they are out of sync, but don't know why, how can we know that syncing them like this won't cause other problems with the data. We're just assuming that every delete on the topic is correct, and that every host missing in the database is correct. I'd like to get to the bottom of the cause, or at least better understand the possibilities before we run this.
That's a good point. Let's consider the assumptions we're making:

- Every `delete` event on `platform.inventory.events` is correct
  - HBI is the only app that emits messages to this topic, and we only do it when we delete a host. In what scenario would there be a `delete` event on that topic that isn't correct?
- Every host missing from the database is correct
  - In other words, there's a `created`/`updated` event on the events topic, and that host isn't in the DB anymore. In that situation, I would have to assume that the record was deleted, but the `delete` event was lost somehow, or that the producer failed to emit it. In that case, I would want this script to create a delete event.

Does that make sense? Please let me know if you've got specific situations in mind; I do want to cover any potential gaps

**Comment by kruai** [comment]
Based on some feedback from @fijshion, we should treat the HBI DB as the source of truth. So, I removed the part that deletes hosts from the DB based on `delete` events. I have some more changes in progress as well, so you might wanna hold off on reviewing more for now

File: deployment.yaml [hunk 2]
```diff
@@ -545,6 +545,54 @@ objects:
...
+          - name: INVENTORY_DB_SSL_MODE
+            value: ${INVENTORY_DB_SSL_MODE}
+          - name: INVENTORY_DB_SSL_CERT
+            value: ${INVENTORY_DB_SSL_CERT}
+          - name: KAFKA_BOOTSTRAP_SERVERS
+            value: ${KAFKA_BOOTSTRAP_HOST}:${KAFKA_BOOTSTRAP_PORT}
+          - name: PAYLOAD_TRACKER_KAFKA_TOPIC
+            value: platform.payload-status
+          - name: PAYLOAD_TRACKER_SERVICE_NAME
+            value: inventory-mq-service
```
Comment (ryandillinfelton) [inline deployment.yaml:568]: We might want to change this, but I don't think anything happening here would be tracked  by the payload tracker anyways

**Review by ryandillinfelton — DISMISSED** [review]
LGTM

**Review by thearifismail — APPROVED** [review]
LGTM
